akka-projection-core/src/main/scala/akka/projection/javadsl/SourceProvider.scala

```diff
@@ -8,18 +8,29 @@ import java.util.Optional
 import java.util.concurrent.CompletionStage
 import java.util.function.Supplier
 
+import akka.projection.MergeableOffset
 import akka.projection.OffsetVerification
-import akka.projection.OffsetVerification.VerificationSuccess
+import akka.projection.ProjectionContext
+import akka.projection.internal.ProjectionContextImpl
 import akka.stream.javadsl.Source
 
-abstract class SourceProvider[Offset, Envelope] {
+trait SourceProvider[Offset, Envelope] {
 
   def source(offset: Supplier[CompletionStage[Optional[Offset]]]): CompletionStage[Source[Envelope, _]]
 
   def extractOffset(envelope: Envelope): Offset
 
-  def verifyOffset(offset: Offset): OffsetVerification = VerificationSuccess
+}
+
+trait VerifiableSourceProvider[Offset, Envelope] extends SourceProvider[Offset, Envelope] {
+
+  def verifyOffset(offset: Offset): OffsetVerification
+
+}
+
+trait MergeableOffsetSourceProvider[Offset <: MergeableOffset[_, _], Envelope] {
 
-  def isOffsetMergeable: Boolean = false
+  private[projection] def groupByKey(envs: java.util.List[ProjectionContextImpl[_, Envelope]])
```

Comment:
`private[projection]` will not work for javadsl if the purpose is to hide it, it will be public